Add documentation for the central data structures in CBMC IR

NlightNFotis opened the pull request:
This PR describes the central GOTO-IR data structures
at a light level of detail (containing some pseudocode).

Please let us know if the level of detail contained in this PR
is adequate, or if there's more that's needed.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add documentation for the central data structures in CBMC IR (goto-programs).

martin-cs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a product type, consisting of:
```
Do we expect readers of our documentation to know what a product type is?

NlightNFotis replied:
I was operating under the assumption that this is developer's documentation, and I expect CBMC developers to be exposed to Programming Language Theory, and thus, know what a product type is.

Is this an unsafe assumption to make?

If yes, I need to rework this and some other documentation that I'm preparing.

martin-cs replied:
Thinking about our undergraduate program, I think the terminology for product, sum and exponential types might get introduced in the third year?  Even then, there are developers who don't have a formal CS background, or have an older one before this kind of terminology was used at undergraduate level.  I wouldn't say safe or unsafe, it's more about how accessible and welcoming we can make the documentation.

Maybe a good starting point is to assume they know C++ terminology, maybe from C++14 or C++17.  So perhaps you could say that a `goto_modelt` is a pair of a `symbol_tablet` and a `goto_functionst`, a `goto_functionst` is a map from function names to `goto_functiont`, etc.

thomasspriggs replied:
"pair" is a much more easily understandable word to use explain this than "product type" in my opinion.

thomasspriggs replied:
"product" is no longer used.

martin-cs commented on doc/architectural/central-data-structures.md:
````diff
@@ -0,0 +1,91 @@
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
+type goto_functiont {
````
Is there a way to link this to the actual code or generated doxygen?

NlightNFotis replied:
I can link to the file using raw markdown, but I don't think it can link to an actual code structure.

I need to look more into doxygen and its linking capabilities.

martin-cs replied:
If you can link it, it might help increase the longevity.  See https://github.com/diffblue/cbmc/pull/7470#issuecomment-1377631947 for my experiences with the lifespan of documentation.  The last thing I want is for you to put in the effort documenting things and then, in 2 years, have to do it all again.

thomasspriggs replied:
We need to link to either the actual code or the generated doxygen output for these things. Duplicating the documentation of the same thing in two places is not sustainable.

NlightNFotis replied:
In the final documentation, `doxygen` is automatically linking class names to the class documentation it produces, so when the class names are written in full they get automatically picked up by it and linked.

However, given that these are central data structures and calcified by now (i.e. not experiencing significant change with regards to the actual components of the class), I'd rather the pseudocode representation stays as a fast-track way for someone to understand the essence of these data structures without the informational load and visual clutter that the class documentation pages in `doxygen` have.

martin-cs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+This is an `irept`. `irep`s are the central data structure that model most entities inside
+CBMC and the assorted tools - effectively a node/map like data structure that forms a hierachical
+tree that ends up modeling graphs like ASTs, CFGs, etc. (This will be further discussed in
+a dedicated page).
```
I am not sure we should commit TODO notes.

NlightNFotis replied:
This was not so much as a TODO, rather, it's a small digression to guide user expectation (that they should expect a dedicated topic on this that they need to look for into the documentation).

Do you think this should leave, or is there a better way to phrase this?

martin-cs replied:
Maybe put in a link when it is available?

TGWDB replied:
I agree on not adding a TODO/promise. Maybe give a really brief overview and link to the source code/doxygen?

NlightNFotis committed:
Change the wording to closer correspond to C++ types and concepts

TGWDB commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,99 @@
+# Central Data Structures
+
+The following is some light technical documentation of the major data structures
+represented in the input transformation to Intermediate Representation (IR) inside
```
"represented" -> "used"
Editing note, try not to use the same word for two different things in one sentence.

TGWDB commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,99 @@
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
```
"use to ..." -> "use for the GOTO-IR (`GOTO` Intermediate Representation)."

TGWDB commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,99 @@
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
```
"IR" not defined. Above the definition was only for "GOTO-IR".

"and effectively it's the following ..." -> "like the following ..."

NlightNFotis replied:
Intermediate representation (IR) is defined in line 4 of the same file.

TGWDB replied:
Not quite, you defined "GOTO-IR", not "IR". I propose change line 4:
"GOTO-IR (GOTO Intermediate Representation)" to something like "GOTO intermediate representation (IR) denoted GOTO-IR"

thomasspriggs replied:
Why say IR when we could just say GOTO?

NlightNFotis replied:
> Why say IR when we could just say GOTO?

Because GOTO is a CBMC concept, and IR is a lot more likely to be a known concept to a computer scientist, so by virtue of connecting the two concepts, we are leveraging people's pre-existing understanding of a general concept to relate it to the internal one.

thomasspriggs replied:
You already explained that in the first paragraph. Repeating it again is redundant.

TGWDB commented on doc/architectural/central-data-structures.md:
````diff
@@ -0,0 +1,99 @@
+}
+```
+
+Of the two types listed above, the `parameters` one should be self-documenting,
+(the values of these are later looked up in the symbol table), so we're going to
+focus next on the type `goto_programt`
````
Stylistically you should describe the parameters in order and consistently (either the type or the name), maybe something like:
"The rest of this section details the `body` and `goto_programt`, the `parameters` are a self-documenting `list` of `identifiers` that are later looked up in the symbol table."

TGWDB commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,99 @@
+  * For illustration purposes, some instruction types are `assign`, `function_call`, `return`,
+    etc.
+* The `instruction` is a statement represented by a `goto_instruction_codet`.
+  * A `goto_instruction_codet` is a `codet` (a data structure broadly representing a statement
+    inside CBMC) that contains the actual GOTO-IR instruction.
```
So a GOTO-IR `modelt` contains GOTO-IR `functiont`s which contains a list of  GOTO-IR `body`s which are GOTO-IR `programt`s that contain **GOTO-IR `instructiont`s** that contain an `instruction` that contains _GOTO-IR `instruction_codet`_ that contain **_GOTO-IR `codet`_** that contain _GOTO-IR `instruction_codet`_ that is a **_GOTO-IR `codet`_** that contains **GOTO-IR `instructiont`**... and this is not at all confusing or circular?

Or have I misread the documentation to this point?

NlightNFotis replied:
I didn't choose the class names 😅

Yes, your understanding is more or less correct. I'll see if I can simplify it further, but I can't promise massive improvements - this is just the description of the status quo.

thomasspriggs replied:
Some of the type names involved seem a little obtuse, which is just why it is so important to document them. Some of the relationships as expressed by @TGWDB are not quite right however. Given that this has been misunderstood by one person already, it suggests that we haven't managed to explain the relationships sufficiently clearly in this documentation.

- A model **contains** a collection of functions.
- Each function **contains** a single "body" which **is** of type `goto_programt`.
- A "body" **contains** an ordered collection of `instructiont`.
- An `instructiont` may optionally **contain** a "code" expression which **is** of type `codet`. `codet` and `goto_instruction_codet` are currently defined to be the same type, but this is not guaranteed to be the case in future.

Note that a function has a single "body" not a collection of them. The fact that a function body is of type `goto_programt` is the especially confusing point, as it is not an entire program! The set of data structures are not circular. There are 4 container ship relations between 5 levels of data.

The `instructiont` class type works a bit like a tagged union. For some values of the instruction type, the `code` member is constrained to be a particular sub-class of `codet`. For others such as the `SKIP` instruction type, the `code` member is expected to be empty (is `ID_nil`). Given that instructions also have other data members `intructiont` is both a sum and product type.

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+
+A `goto_modelt` is effectively a product type, consisting of:
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
```
I suggest describing this as "A collection of GOTO functions" rather than a list, in order to avoid readers potentially conflating this with the linked list data structure. The exact data structure used within the class is a map, but this is not relevant to the point here in the documentation.

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+A `goto_modelt` is effectively a product type, consisting of:
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
+* A symbol table containing symbol references for the symbols contained in the
```
I would say that the symbol table contains the symbols referred to by symbol expressions. Symbol expressions are found both in goto functions and in the values of other symbols. The primary purpose of the symbol table is to store the symbols not references to symbols. So the corrected sentence should be "A symbol table containing a collection of symbols, which are referenced within the GOTO functions."

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
```
Can we get rid of the terminology "product type" here as well? We could explain the entirety of what a goto function is just by saying "A `goto_functiont` is a pairing of a `goto_programt` body and a collection of parameters." I am not sure the pseudocode adds any value in this case.

thomasspriggs commented on doc/architectural/central-data-structures.md:
````diff
@@ -0,0 +1,91 @@
+}
+```
+
+Of the two types listed above, the `parameters` one should be self-documenting,
````
I disagree with your assertion that "parameters" is self documenting. We should explain that a parameter identifier can be looked up in the symbol table (they do need to exist in the symbol table right?!). This may be obvious for an experienced CBMC developer but we should spell it out for those who are new to cbmc. This is important as an inexperienced developer could do something like constructing a goto function which has parameters which don't exist in the symbol table. We should also explain what information the parameter symbol should contain. Does the parameter symbol have the value set to be an instance of `code_typet::parametert`?

jimgrundy replied:
+1

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+type goto_instructiont {
+    instruction_type instr_type
+    instruction      statement
+    guard            boolean_expr
```
Is `source_location` missing from this list?

jimgrundy replied:
+1

tautschnig replied:
Among others? There are indeed various other data members, not all of which we should keep in there long-term, but if you are seeking to document the current state then all of them should be in here?!

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,100 @@
+Another important data structure that needs to be discussed at this point is
+`source_locationt`.
+
+This is an `irept`. `irep`s are the central data structure that model most entities inside
```
Is the fact that `source_locationt` inherits from `irept` relevant to this part of the documentation? I would consider this to be a low-level implementation detail and omit it from the high level overview.

tautschnig replied:
Agreed, but it might be important to cover `irept` by itself?

thomasspriggs replied:
I have separated the explanation of `irept` itself and added further details about it.

thomasspriggs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,100 @@
+
+It might be possible that a specific source location might point to a CBMC instrumentation
+primitive (which might be reported as a `built-in-addition`) or there might even be no-source
+location (because it might be part of harnesses generated as an example, that have no presence
```
:pick: "harness" is not well a well defined at this point in the documentation. We use it to mean the cbmc generated code surrounding the entry point specified to cbmc. This generated harness initialises global values, and the arguments for the entry points parameters. However my understanding is that there are users who write their own "harness" functions in C around the function they are verifying. So in this context there would be 2 harnesses when cbmc runs. The cbmc harness which calls the user written harness which calls the function the user is verifying.

jimgrundy replied:
I kind of object to the notion that we allow code with no location. We don't just generate code out of thin air for no reason. I would think that every function that can generate code should be required to take a location parameter, and I'd like to see us add an assertion that it is non-empty.

We just went through this with the function contracts and had to go all the way back to the front end to make sure that the necessary location information information was being piped down to the point where code was being generated. But, the result is code that is user debuggable.

thomasspriggs reviewed (changes requested):
The documentation in this PR provides a reasonable summary of each of the individual classes involved. However I think this is still missing an overview of the way they relate to each other. This should be separate from the sections documenting the individual classes. It would be nice to have these relationships shown as a diagram.

thomasspriggs committed:
Consistently refer to GOTO as GOTO

To avoid introducing the new term `GOTO-IR` which is not widely used
within the codebase.

thomasspriggs committed:
Correct tense of `used` to `use`.

thomasspriggs committed:
Clarify what `goto_modelt` is used for

Is was not previously clear that `goto_modelt` was used as a storage
class.

thomasspriggs committed:
Replace "main" with "top level" in goto model explaination

Main suggests that there are other entities which hold goto, rather than
expressing that all the other data structures are contained within the
model.

thomasspriggs committed:
Add goto model entity relationship diagram

thomasspriggs committed:
Make central-data-structures a subpage of the architecture page

In order to prevent it being included at the root level of the
documentation site.

thomasspriggs commented:
FYI - For the benefit of those who may be watching this PR, the work to address the comments on this has been handed over to myself from Fotis.

thomasspriggs committed:
Add mermaid diagram support to documentation site build

So that we can add a relationship diagram to the central data structures
documentation page.

tautschnig commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,145 @@
+
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
```
Perhaps say "A collection" (the precise data type is immaterial here, but people might wrongly be led to think it actually is a list).

thomasspriggs replied:
Its actually a mapping from the function identifier to the GOTO function.

tautschnig commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,145 @@
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
```
What is a "symbol reference?" Also, shouldn't the abstraction level be kept consistent and this just be "A collection of the symbols used in GOTO functions."

thomasspriggs replied:
I have now reworked the explanation around symbols / the symbol_table.

tautschnig commented on doc/architectural/central-data-structures.md:
````diff
@@ -0,0 +1,145 @@
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
+
+In pseudocode, the type looks this:
+
+```js
+type goto_modelt {
+  type goto_functionst = list<goto_functiont>
+  type symbol_tablet = map<identifier, symbolt>
+}
+```
````
Does this add value?

tautschnig commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,145 @@
+a more elaborate explanation in the next section.
+
+The `parameters` subcomponent is a list of identifiers that are to be looked-up
+in the symbol-table for their values.
```
Is "values" the most appropriate term here?

thomasspriggs replied:
"value" is a word which is a little too general / overloaded in this context. I have gone for "definition" instead for the moment.

tautschnig commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,145 @@
+
+## goto_programt
+
+A `goto_programt` is a list of GOTO instructions. In pseudocode, it would
+look like `type goto_programt = list<goto_instructiont>`.
```
What's the deal with pseudocode? Is that sentence adding any value?

thomasspriggs replied:
I have removed this particular example of pseudocode.

thomasspriggs committed:
Correct "list of GOTO functions" to "collection"

Because the data structure used is not a list and because we should
avoid being specific here in order to maintain a high level of
abstraction.

thomasspriggs committed:
Reword description of the symbol table

Because the symbol table contains the symbols; not symbol expressions.

thomasspriggs committed:
Remove pseudocode from the description of `goto_programt`

thomasspriggs committed:
Reword description of goto program

To draw attention to the link to the `goto_programt` classes'
documentation. Note that class names are automatically linked in
generating the website.

thomasspriggs committed:
Add explaination of heirarchy of goto model classes

This existence of an abstract interface was previously mentioned but not
properly explained or justified.

thomasspriggs committed:
Show that `goto_functionst` is a map not a list

thomasspriggs committed:
Reword explaination of symbol table

In order to avoid the overloaded word "value".

thomasspriggs committed:
Add links to `symbolt` and `symbol_exprt` classes

In order to make it easier to find the documentation concerning these
classes.

thomasspriggs committed:
Note optional nature of goto program bodies

thomasspriggs committed:
Separate the documentation of the `irept` class

Because `irept` itself should be considered to live at a separate level
of abstraction from the other classes.

thomasspriggs committed:
Provide more extensive overview of `irept` itself

tautschnig reviewed (changes requested):
It only occurred to me (thanks to a reminder by @zhassan-aws) that we have https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/README.md. Can I please ask that we do a decent editor's job and avoid having similar information in multiple places?

thomasspriggs commented:
> It only occurred to me (thanks to a reminder by @zhassan-aws) that we have https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/README.md. Can I please ask that we do a decent editor's job and avoid having similar information in multiple places?

The documentation being added in this PR is intended to be an architectural overview which explains how the different data structures fit together and where the detailed documentation for each of them can be found. Of a necessity, a less detailed overview will include information which is explained in more detail in the documentation which is specific to each data structure. It is also expected that this architectural overview will include points which are not directly relevant to goto-programs, such as the section we have now added on `irept`.

I can update the documentation added here to link to `goto-programs/README.md`

tautschnig commented:
> I can update the documentation added here to link to `goto-programs/README.md`

Yes, please make sure there is cross-referencing and also no contradicting information (redundant information likely is unavoidable, but that's ok).

thomasspriggs committed:
Add link to goto program documentation from central data structures

thomasspriggs committed:
Replace relative link with web link

In order to fix doxygen error.

tautschnig commented on doc/architectural/central-data-structures.md:
```diff
@@ -111,7 +111,7 @@ A  goto program  is a sequence of GOTO instructions (`goto_instructiont`). For
 [`goto_program.h`](../../src/goto-programs/goto_program.h). For further details
 of the life cycle of goto programs see
-[`src/goto-programs/README.md`](../../src/goto-programs/README.md).
+[`src/goto-programs/README.md`](https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/README.md).
```
What will this be rendered as on diffblue.github.io?!

thomasspriggs replied:
This link follows the same format as is found here - https://github.com/diffblue/cbmc/blob/3f578e18a897ac4f47ba7cff8b0d9c04c9bb799c/doc/architectural/compilation-and-development.md?plain=1#L17
this example is rendered to https://diffblue.github.io/cbmc/compilation-and-development.html

The original link I used was causing the check-doxygen job to fail with the error - `/home/runner/work/cbmc/cbmc/doc/architectural/central-data-structures.md:120: warning: unexpected token TK_EOF as the argument of ref`

tautschnig replied:
Well, but `COMPILING.md` isn't actually rendered on diffblue.github.io, while `goto-programs/README.md` is (the latter as https://diffblue.github.io/cbmc/group__goto-programs.html). It would be great to find a way so that people aren't sent back&forth between Doxygen-rendered content and GitHub rendered pages.

NlightNFotis reviewed (commented):
LGTM - Approved.

(Cannot give green tick because Github UI doesn't allow the person who raised a PR to approve or request changes).

tautschnig reviewed (approved):
Approving with the agreement that #7497 will take care of cleanup.
